### PR TITLE
[AArch64] Let patterns for NEON instructions check runtime mode.

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
This helps identify any failures where the compiler might otherwise
silently emit instructions that are not valid for the given runtime
mode. We can probably do a similar thing for HasSVE predicates.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 1]
```diff
@@ -5934,7 +5929,7 @@ def : Pat<(v2f64 (AArch64frsqrts (v2f64 FPR128:$Rn), (v2f64 FPR128:$Rm))),
 // Some float -> int -> float conversion patterns for which we want to keep the
 // int values in FP registers using the corresponding NEON instructions to
 // avoid more costly int <-> fp register transfers.
-let Predicates = [HasNEON] in {
+let Predicates = [HasNEONorSME] in {
```
Comment (efriedma-quic) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:5932]: Comment on the closing brace also needs to be updated.

**Review by efriedma-quic — COMMENTED** [review]
If I'm understanding correctly, this makes "HasNEONorSME" mean "this target supports NEON encodings", and "HasNEON" means "this target supports NEON encodings, and the current configuration of the SME registers allows executing the full NEON instruction set".

I don't think I like these names... can we rename them in a followup?  Maybe use the same names as the Subtarget methods?

That said, the approach here makes sense.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 2]
```diff
@@ -234,12 +234,9 @@ def HasSMEF16F16orSMEF8F16
...
 
-// A subset of NEON instructions are legal in Streaming SVE execution mode,
-// they should be enabled if either has been specified.
-def HasNEONorSME
-    : Predicate<"Subtarget->hasNEON() || Subtarget->hasSME()">,
-                AssemblerPredicateWithAll<(any_of FeatureNEON, FeatureSME),
-                "neon or sme">;
+// A subset of NEON instructions are legal in Streaming SVE execution mode.
+def HasNEONorSME     : Predicate<"Subtarget->hasNEON()">,
+                                 AssemblerPredicateWithAll<(any_of FeatureNEON), "neon">;
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.td:239]: `HasNEONorSME` but the predicate doesn't check for SME?

**Comment by paulwalker-arm** [comment]
Personally I don't see value in a mass rename of `HasNEON` because the runtime mode checking only applies to isel and not assembly.  I see the overriding of the predicate as just a convenient way to restrict only isel for the common case of most all NEON instructions not being safe to execute in streaming mode.

I guess I'm worried we'll introduce complexity where it's not needed because the modal behaviour is mostly handled before isel and if we get this far we're already toast.  The rational for change HasNEON is mainly to fail in a nicer way.

That said, I do think we're at the point where `HasNEONOrSME` doesn't represent what we mean (i.e. it's hard to tell but I don't think it's the case that an implementation with SME but not NEON must have these instructions?) so perhaps `HasNEONAndStreamingSafe` is a better name to show the separate feature (i.e. Inst) and runtime (isel) requirements whilst also implying blocks protected by `HasNEON` are not streaming safe.

**Review by paulwalker-arm — APPROVED** [review]
One observation but otherwise this looks good to me.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.td [hunk 3]
```diff
@@ -6016,7 +6013,7 @@ def : Pat<(v2f64 (AArch64frsqrts (v2f64 FPR128:$Rn), (v2f64 FPR128:$Rm))),
 // Some float -> int -> float conversion patterns for which we want to keep the
 // int values in FP registers using the corresponding NEON instructions to
 // avoid more costly int <-> fp register transfers.
-let Predicates = [HasNEON] in {
+let Predicates = [HasNEONandIsStreamingSafe] in {
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply llvm/lib/Target/AArch64/AArch64InstrInfo.td:6016]: Good point, done!

**Pull request merged**